docs/proposals/20190919-machinepool-api.md

```diff
@@ -0,0 +1,419 @@
+
+#### States and Transitions
+
+##### Pending
+
+``` go
+// MachinePoolPhasePending is the first state a MachinePool is assigned by
+// Cluster API MachinePool controller after being created.
+MachinePoolPhasePending = MachinePoolPhase("pending")
+```
+
+###### Transition Conditions
+
+- MachinePool.Phase is empty
+
+###### Expectations
+
+- When MachinePool.Spec.Template.Spec.Bootstrap.Data is:
+  - <nil>, expect the field to be set by an external controller.
+  - “” (empty string), expect the bootstrap step to be ignored.
+  - “...” (populated by user or from the bootstrap provider), expect the contents to be used by a
+    bootstrap or infra provider.
+- When MachinePool.Spec.Template.Spec.InfrastructureRef is:
+  - <nil> or not found, expect InfrastructureRef will be set/found during subsequent requeue.
+  - “...” (populated by user) and found, expect the infrastructure provider is waiting for bootstrap
+    data to be ready.
+  - Found, expect InfrastructureRef to reference an object such as GoogleManagedInstanceGroup,
+    AWSAutoScaleGroup, or AzureVMSS.
+
+##### Provisioning
+
+``` go
+// MachinePoolPhaseProvisioning is the state when the
+// MachinePool infrastructure is being created.
+MachinePoolPhaseProvisioning = MachinePoolPhase("provisioning")
+```
+
+###### Transition Conditions
+
+- MachinePool.Spec.Template.Spec.Bootstrap.ConfigRef -> Status.Ready is true
+- MachinePool.Spec.Template.Spec.Bootstrap.Data is not <nil>
+
+###### Expectations
+
+- MachinePool’s infrastructure to be in the process of being provisioned.
+
+##### Provisioned
+
+``` go
+// MachinePoolPhaseProvisioned is the state when its
+// infrastructure has been created and configured.
+MachinePoolPhaseProvisioned = MachinePoolPhase("provisioned")
+```
+
+###### Transition Conditions
+
+- MachinePool.Spec.Template.Spec.InfrastructureRef -> Status.Ready is true
+- MachinePool.Status.Replicas is synced from MachinePool.Spec.Template.Spec.InfrastructureRef -> Status.Replicas
+
+###### Expectations
+
+- MachinePool’s infrastructure has been created and the compute resources are configured.
+
+##### Running
+
+``` go
+// MachinePoolPhaseRunning is the MachinePool state when it has
+// become a set of Kubernetes Nodes in a Ready state.
+MachinePoolPhaseRunning = MachinePoolPhase("running")
+```
+
+###### Transition Conditions
+
+- Number of Kubernetes Nodes in a Ready state matching MachinePool.Spec.Selector equal MachinePool.Status.Replicas.
+
+###### Expectations
+
+- MachinePool controller should set MachinePool.Status.NodeRefs.
+
+##### Deleting
+
+``` go
+// MachinePoolPhaseDeleting is the MachinePool state when a delete
+// request has been sent to the API Server,
+// but its infrastructure has not yet been fully deleted.
+MachinePoolPhaseDeleting = MachinePoolPhase("deleting")
+```
+
+###### Transition Conditions
+
+- MachinePool.ObjectMeta.DeletionTimestamp is not <nil>
+
+###### Expectations
+
+- Nodes associated with MachinePool should be cordoned and drained.
```

Comment:
You mentioned above that cordon & drain is out of scope for this initial work?

Reply:
That's correct. I removed this line. I'd like to be able to add cordon and drain support after we get initial MachinePool support in.